### PR TITLE
ENH: Support for BIDS event files

### PR DESCRIPTION
**Pull request by TheChymera**
SpecifyModel now supports BIDS event files as inputs.

## Acknowledgment

- [x] \(Mandatory\) I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

**Comment by effigies** [comment]
Just a heads up that we're working on putting all of this functionality into [pybids](https://github.com/bids-standard/pybids/) and [FitLins](https://github.com/poldracklab/fitlins) (see [`LoadBIDSModel`](https://github.com/poldracklab/fitlins/blob/f1d9a84d86a9592cee08ac8793dff17c16248fe1/fitlins/interfaces/bids.py#L134-L309), which will produce outputs similar to `SpecifyModel` and we'll need to make adapted FSL, SPM and AFNI interfaces to handle the differences). The interfaces will be upstreamed to nipype when stabilized.

That said, this is a pretty light-weight addition that will make adapting existing pipelines to handle BIDS-formatted inputs easy, especially in the absence of BIDS stats models, which is still a draft spec.

Can you merge `master` to fix the failing test?

File: nipype/algorithms/modelgen.py [hunk 1]
```diff
@@ -144,6 +144,61 @@ def scale_timings(timelist, input_units, output_units, time_repetition):
     timelist = [np.max([0., _scalefactor * t]) for t in timelist]
     return timelist
 
+def bids_gen_info(bids_event_files,
+    condition_column='trial_type',
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]: The space before each parameter should align with the open paren:
```suggestion
                  condition_column='trial_type',
```

File: nipype/algorithms/modelgen.py [hunk 2]
```diff
@@ -144,6 +144,61 @@ def scale_timings(timelist, input_units, output_units, time_repetition):
...
+    amplitude_column=None,
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]:
```suggestion
                  amplitude_column=None,
```

File: nipype/algorithms/modelgen.py [hunk 3]
```diff
@@ -144,6 +144,61 @@ def scale_timings(timelist, input_units, output_units, time_repetition):
...
+    time_repetition=False,
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]:
```suggestion
                  time_repetition=False):
```

File: nipype/algorithms/modelgen.py [hunk 4]
```diff
@@ -144,6 +144,61 @@ def scale_timings(timelist, input_units, output_units, time_repetition):
...
+    ):
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]:
```suggestion
```

File: nipype/algorithms/modelgen.py [hunk 5]
```diff
@@ -144,6 +144,61 @@ def scale_timings(timelist, input_units, output_units, time_repetition):
...
+    -------
+
+    list of Bunch
+    """
+    info = []
+    for bids_event_file in bids_event_files:
+        with open(bids_event_file) as f:
+            f_events = csv.DictReader(f, skipinitialspace=True, delimiter='\t')
+            events = [{k: v for k, v in row.items()} for row in f_events]
+            conditions = list(set([i[condition_column] for i in events]))
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]: You can release the file handle as soon as it has been fully read.

```suggestion
        conditions = list(set([i[condition_column] for i in events]))
```

File: nipype/algorithms/modelgen.py [hunk 6]
```diff
@@ -144,6 +144,61 @@ def scale_timings(timelist, input_units, output_units, time_repetition):
...
+                    runinfo.conditions.append(condition)
+                else:
+                    runinfo.conditions.append('e0')
+                runinfo.onsets.append(onsets)
+                runinfo.durations.append(durations)
+                try:
+                    amplitudes = [float(i[amplitude_column]) for i in selected_events]
+                    runinfo.amplitudes.append(amplitudes)
+                except KeyError:
+                    runinfo.amplitudes.append([1]*len(onsets))
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]:
```suggestion
                    runinfo.amplitudes.append([1] * len(onsets))
```

File: nipype/algorithms/modelgen.py [hunk 7]
```diff
@@ -144,6 +144,61 @@ def scale_timings(timelist, input_units, output_units, time_repetition):
...
+            events = [{k: v for k, v in row.items()} for row in f_events]
+            conditions = list(set([i[condition_column] for i in events]))
+            runinfo = Bunch(conditions=[], onsets=[], durations=[], amplitudes=[])
+            for condition in conditions:
+                selected_events = [i for i in events if i[condition_column]==condition]
+                onsets = [float(i['onset']) for i in selected_events]
+                durations = [float(i['duration']) for i in selected_events]
+                if time_repetition:
+                    decimals = math.ceil(-math.log10(time_repetition))
+                    onsets = [round(i,decimals) for i in onsets]
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]:
```suggestion
                    onsets = [round(i, decimals) for i in onsets]
```

File: nipype/algorithms/modelgen.py [hunk 8]
```diff
@@ -144,6 +144,61 @@ def scale_timings(timelist, input_units, output_units, time_repetition):
...
+                    durations = [round(i,decimals) for i in durations]
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]:
```suggestion
                    durations = [round(i, decimals) for i in durations]
```

File: nipype/algorithms/modelgen.py (hunk 3 above)
Comment (TheChymera) [reply nipype/algorithms/modelgen.py]: this is better, because it allows you to add parameters without editing the lines belonging to other parameters.

Comment (effigies) [reply nipype/algorithms/modelgen.py]: Okay.

**Comment by effigies** [comment]
@TheChymera Can you merge/rebase current master?

**Pull request merged**